components/codec/src/byte.rs

```diff
@@ -352,6 +353,35 @@ impl MemComparableCodecHelper for DescendingMemComparableCodecHelper {
     }
 }
 
+pub trait BufferByteEncoder: BufferNumberEncoder {
+    fn write_bytes(&mut self, bs: &[u8], desc: bool) -> Result<()> {
+        let len = MemComparableByteCodec::encoded_len(bs.len());
+        let buf = unsafe { self.bytes_mut(len) };
+        if buf.len() < len {
+            return Err(Error::BufferTooSmall);
+        }
+        if desc {
+            MemComparableByteCodec::encode_all_desc(bs, buf);
+        } else {
+            MemComparableByteCodec::encode_all(bs, buf);
+        }
+        unsafe {
+            self.advance_mut(len);
+        }
+        Ok(())
+    }
+
+    /// Joins bytes with its length into a byte slice. It is more
+    /// efficient in both space and time compared to `encode_bytes`. Note that the encoded
+    /// result is not memcomparable.
+    fn write_compact_bytes(&mut self, data: &[u8]) -> Result<()> {
```

Comment:
Personally I would like to create a new encoder for it, maybe call it `CompactByteEncoder`, or more precise one `LengthPrefixByteEncoder`.